base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -821,24 +828,69 @@ where T: BlockchainBackend
             }
             hash = b.header.prev_hash.clone();
             height -= 1;
-            reorg_chain.push_front(b);
+            fork_chain.push_front(b);
         }
-
-        Ok(reorg_chain)
-    }
-
-    fn build_orphan_tree(&self, _reorg_chain: VecDeque<Block>) -> Result<(), ChainStorageError> {
-        // TODO - Build an return a tree for later blocks leading to new_block
-        Ok(())
+        // Check if the constructed fork chain is connected to the main chain.
+        let fork_start_header = fork_chain
+            .front()
+            .expect("The new orphan block should be in the queue")
+            .header
+            .clone();
+        if let Ok(header) = self.fetch_header_with_block_hash(fork_start_header.prev_hash) {
+            if header.height + 1 == fork_start_header.height {
+                return Ok(fork_chain);
+            }
+        }
+        Ok(VecDeque::new())
+    }
+
+    /// Try to find all orphan chain tips that originate from the current orphan parent block.
+    fn find_orphan_chain_tips(&self, parent_height: u64, parent_hash: BlockHash) -> Vec<BlockHash> {
+        let mut tip_hashes = Vec::<BlockHash>::new();
+        self.db
+            .for_each_orphan(|pair| {
+                let (_, block) = pair.unwrap();
+                if (block.header.prev_hash == parent_hash) && (block.header.height == parent_height + 1) {
+                    let next_parent_hash = block.hash();
+                    let mut orphan_chain_tips =
+                        self.find_orphan_chain_tips(block.header.height, next_parent_hash.clone());
+                    if orphan_chain_tips.len() > 0 {
+                        tip_hashes.append(&mut orphan_chain_tips);
+                    } else {
+                        tip_hashes.push(next_parent_hash);
+                    }
+                }
+            })
+            .expect("Unexpected result for database query");
+        if tip_hashes.len() == 0 {
+            // No chain tips found, then parent must be the tip.
+            tip_hashes.push(parent_hash);
+        }
+        tip_hashes
     }
 
-    fn find_longer_chain(&self, _tree: ()) -> Option<Vec<Block>> {
-        // TODO - Check tree vs main chain for great accum difficulty
-        None
+    /// Find and return the orphan chain tip with the highest accumulated difficulty.
+    fn find_strongest_orphan_tip(
```

Comment:
I think we could optimise by doing this during the `.for_each_orphan` loop above; but it's written nice and clearly right now, so I reckon we can leave it as is and revisit if performance is on the slower side.

Reply:
I agree, it looks like it can be integrated into the `for_each_orphan`. Let's revisit it again in the future.